diff --git configs/socfpga_de0_nano_soc_defconfig configs/socfpga_de0_nano_soc_defconfig
index 65c1197..2691e50 100644
--- configs/socfpga_de0_nano_soc_defconfig
+++ configs/socfpga_de0_nano_soc_defconfig
@@ -21,3 +21,7 @@ CONFIG_DESIGNWARE_SPI=y
 CONFIG_DM_MMC=y
 CONFIG_USB=y
 CONFIG_DM_USB=y
+CONFIG_OF_EMBED=y
+CONFIG_MENU=y
+CONFIG_MENU_SHOW=y
+CONFIG_CMD_BOOTMENU=y
diff --git include/configs/socfpga_de0_nano_soc.h include/configs/socfpga_de0_nano_soc.h
index cbc7396..7381008 100644
--- include/configs/socfpga_de0_nano_soc.h
+++ include/configs/socfpga_de0_nano_soc.h
@@ -34,10 +34,31 @@
 #define PHYS_SDRAM_1_SIZE		0x40000000	/* 1GiB */
 
 /* Booting Linux */
+#define CONFIG_PREBOOT \
+	"if mmc rescan; then " \
+		"if run loadbootenv; then " \
+			"echo Loaded environment from ${bootenv};" \
+			"run importbootenv;" \
+		"fi;" \
+	"fi;"
+#define CONFIG_MENU
+#define CONFIG_MENU_SHOW
+#define CONFIG_CMD_BOOTMENU
+
 #define CONFIG_BOOTDELAY	3
 #define CONFIG_BOOTFILE		"fitImage"
 #define CONFIG_BOOTARGS		"console=ttyS0," __stringify(CONFIG_BAUDRATE)
-#define CONFIG_BOOTCOMMAND	"run mmcload; run mmcboot"
+#define CONFIG_BOOTCOMMAND     \
+	"if mmc rescan; then " \
+		"if test -n $uenvcmd; then " \
+			"echo Running uenvcmd ...;" \
+			"run uenvcmd;" \
+		"fi;" \
+		"if run loadbootscript; then " \
+			"run bootscript; " \
+		"fi; " \
+	"fi;" \
+        "run mmcload; run mmcboot"
 #define CONFIG_LOADADDR		0x01000000
 #define CONFIG_SYS_LOAD_ADDR	CONFIG_LOADADDR
 
@@ -52,10 +73,13 @@
 /* Extra Environment */
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	"loadaddr= " __stringify(CONFIG_SYS_LOAD_ADDR) "\0" \
+	"loadbootenv_addr= " __stringify(CONFIG_SYS_LOAD_ADDR) "\0" \
+	"kernel_addr_r= " __stringify(CONFIG_SYS_LOAD_ADDR) "\0" \
 	"ramboot=setenv bootargs " CONFIG_BOOTARGS ";" \
 		"bootm ${loadaddr} - ${fdt_addr}\0" \
 	"bootimage=zImage\0" \
 	"fdt_addr=100\0" \
+	"fdt_addr_r=100\0"        \
 	"fdtimage=socfpga.dtb\0" \
 	"bootm ${loadaddr} - ${fdt_addr}\0" \
 	"mmcroot=/dev/mmcblk0p2\0" \
@@ -65,6 +89,15 @@
 	"mmcload=mmc rescan;" \
 		"load mmc 0:1 ${loadaddr} ${bootimage};" \
 		"load mmc 0:1 ${fdt_addr} ${fdtimage}\0" \
+        "fpgadata=0x02000000\0" \
+	"bootenv=uEnv.txt\0" \
+        "config_name=" CONFIG_SYS_CONFIG_NAME "\0" \
+	"loadbootenv=load mmc 0 ${loadaddr} ${bootenv}\0" \
+	"importbootenv=echo Importing environment from mmc ...;" \
+		"env import -t $loadaddr $filesize\0" \
+        "loadbootscript=load mmc 0 ${loadaddr} boot.scr\0" \
+        "bootscript=echo Running bootscript from mmc ...;" \
+                "source ${loadaddr}\0" \
 
 /* The rest of the configuration is shared */
 #include <configs/socfpga_common.h>
